<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Maize Pangenome Variant Effect Prediction Tool</title>
    <script src="https://d3js.org/d3.v5.min.js"></script>
    <script src="./js/main.js"></script>
    <script src="./js/dom.js"></script>
    <script src="./js/support.js"></script>
    <script src="./js/genome.js"></script>
    <script src="./js/pan.js"></script>

    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.2.0/css/all.css">
    <link rel="stylesheet" href="./css/main.css">

    <script>
        if(getQueryStringValue("id"))
        {
            main_id = getQueryStringValue("id");
        }

        if(getQueryStringValue("option"))
        {
            main_option = getQueryStringValue("option");
        }

        if (main_id.includes("_T")) {
            let parts = main_id.split("_");
            gene_model = parts[0];
            transcript = main_id;
            protein = main_id.replace("_T", "_P");
        } else if (main_id.includes("_P")) {
            let parts = main_id.split("_");
            gene_model = parts[0];
            transcript = main_id.replace("_P", "_T");
            protein = main_id;
        } else {
            gene_model = main_id;
            can_flag = true;
            transcript = main_id + "_T001";
            protein = main_id + "_P001";
        }
    </script>

</head>

<body>
 <div id="loading-icon" style="display: none;">
     <img src="./img/loading.gif" width="100px" alt="Loading..."/>
 </div>
 <div class="main-container">
    <div class="header">
        MaizeGDB Variant Effect Tool
    </div>

    <div class="tabs">
        <div>
            <button class="tab" id="summary_button" onclick="showContent('summary');highlightTab('summary_button')">Gene Summary</button>
            <button class="tab" id="b73_button" onclick="showContent('b73');highlightTab('b73_button')">Variant effects in B73</button>
            <button class="tab" id="pan_button" onclick="showContent('pan-genome');highlightTab('pan_button')">Variant effects across the pangenome</button>
            <button class="tab" id="search_button" onclick="showContent('search');highlightTab('search_button')">Search</button>
        </div>
        <button class="tab" id="help_button" onclick="showContent('help');highlightTab('help_button')">Help</button> <!-- Help tab on far right -->
    </div>

    <div id="summary" class="content">
        <div id="gene-container" class="gene-model-container">
            <div class="column">
            <div class="gene-model-title">Gene Summary</div>
                <div id=gene-summary class="text-content">
                    <ul class="gene-model-features">
                        <li>Genome: <a target="_blank" href="https://www.maizegdb.org/genome/assembly/Zm-B73-REFERENCE-NAM-5.0">Zm-B73-REFERENCE-NAM-5.0</a></li>
                        <li>Gene model: <span id="summary_gm"></span></li>
                        <li>Transcript: <span id="summary_transcript"></span></li>
                        <li>Canonical: <span id="summary_canonical"></span></li>
                        <li>Gene Symbol: <span id="summary_gs"></span></li>
                        <li>Gene Name: <span id="summary_gn"></span></li>
                    </ul>
                </div>
            <div class="gene-model-title">Protein Summary</div>
                <div id=protein-summary class="text-content">
                    <ul class="gene-model-features">
                        <li>Protein: <span id="summary_protein"></span></li>
                        <li>Protein Length: <span id="summary_pl"></span></li>
                        <li>Pan MSA length: <span id="summary_msa"></span></li>
                        <li>UniProt: <span id="summary_uniprot"></span></li>
                        <li>UniProt Description: <span id="summary_desc"></span></li>
                        <li>3D Structure: <span id="summary_ss"></span></li>
                    </ul>
                    <br><span id="summary_download"></span>
                </div>
            </div>
            <div class="column">
               <div class="gene-model-title">GWAS-based traits</div>
               <div id=phenotype-summary class="text-content">

               </div>
           </div>
            </div>
    </div>

    <div id="b73" class="content">
        <div class="sectionHeader">PFAM Domains</div>
        <div id="pfam-wrap">
            <div id="pfamNumberLine" class="numberLine"></div>
            <div id="pfamGeneModel" class="geneModel"></div>
        </div>

        <div class="sectionHeader">Secondary Structure</div>
        <div id="dssp">
            <canvas id="proteinStructure" width="1400px" height="80px"></canvas>
        </div>

        <div class="sectionHeader">Variant Effects of full protein (heatmap)</div>
        <div id="heatNumberLine" class="numberLine"></div>
        <div class="heatmap-container">
            <div id="full-heatmap">
                <div id="highlight-box"></div>
            </div>
        </div>

        <div class="slider-container" id="slider-container">
            <span id="slider"></span>
            <span id="slider-value">1</span>
        </div>

        <div class="sectionHeader">Variant Effects of zoomed in region (heatmap)</div>
        <div id="zoomNumberLine" class="numberLine"></div>
        <div id="heatmap-container-zoom" class="heatmap-container-zoom">
            <div id="zoomed-heatmap"></div>
        </div>
        <div id="zoomWTLine" class="numberLine"></div>
    </div>

    <div id="pan-genome" class="content">
        <div class="sectionHeader">PFAM Domains</div>
        <div id="pfam-wrap-pan">
            <div id="pfamNumberLine-pan" class="numberLine"></div>
            <div id="pfamGeneModel-pan" class="geneModel"></div>
        </div>

        <div class="sectionHeader">Secondary Structure</div>
        <div id="dssp-pan">
            <canvas id="proteinStructure-pan" width="1400px" height="80px"></canvas>
        </div>

        <div class="sectionHeader" id="heatheader-pan">Heatmap of full protein</div>
        <div id="heatNumberLine-pan"-pan class="numberLine"></div>
        <div id="colorBox-pan"></div>
        <div class="heatmap-container-pan">
            <div id="full-heatmap-pan">
                <div id="highlight-box-pan"></div>
            </div>
        </div>

        <div class="slider-container" id="slider-container-pan">
            <span id="slider-pan"></span>
            <span id="slider-value-pan">1</span>
        </div>
        <div class="sectionHeader">Heatmap of zoomed in region</div>
        <div id="zoomNumberLine-pan" class="numberLine"></div>
        <div id="heatmap-container-zoom-pan" class="heatmap-container-zoom-pan">
            <div id="zoomed-heatmap-pan" ></div>
        </div>
        <div id="zoomWTLine-pan" class="numberLine"></div>

    </div>

    <div id="search" class="content">
        <div class="gene-model-container-pan">
            <div class="gene-model-title">Variant Effects Search</div>
            <form action="./index.html" method="get">
                <div class="wrap">
                    <div class="search">
                        <input name="id" id="p1" type="text" class="searchTerm" placeholder="Enter gene, protein or transcript id">
                        <button type="submit" class="searchButton">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>
                    <div class="example">Examples: <a href="javascript:changeExample('Zm00001eb057210')">Zm00001eb057210</a>, <a href="javascript:changeExample('Zm00001eb057210_P001')">Zm00001eb057210_P001</a>, <a href="javascript:changeExample('Zm00001eb057210_T001')">Zm00001eb057210_T001</a></div>
                    <br><br>
                    <div class="radio-options">
                        <label>
                            <input type="radio" name="option" value="both" checked> Show both B73 and pangenome views
                        </label>
                        <label>
                            <input type="radio" name="option" value="b73"> Show B73 view only
                        </label>
                        <label>
                            <input type="radio" name="option" value="pan"> Show pangenome view only
                        </label>
                   </div>

                    <div id="search_error" class="text-content"></div>
                        <div class="text-content">
                        <br>Summary: This tool visualizes the potential effects of missense variants on proteins, showcasing the impact of amino acid substitutions for each residue. Using the <a target="_blank" href="https://github.com/facebookresearch/esm">ESM protein language model</a> through the <a target="_blank" href="https://github.com/ntranoslab/esm-variants">esm-variants tool</a>, it calculates variant effect scores based on the log-likelihood ratio between the variant and its wild-type. Scores above -7 indicate benign outcomes, while scores below -7 suggest possible phenotypic effects. A score of 0 denotes when the variant and wild-type are the same. Our heatmap representation range from blue (benign effects) to green / yellow (mild effects) to red (phenotypic effect). Additionally, Pfam domains and predicted secondary structures are included to highlight and provide context to the protein's functional domains.

                        <br><br>This tool has two views:

                        <br><br><b>Variant effects in B73</b>: This view shows the predicted impact of all possible amino acid substitutions for proteins in the B73 genome. The views are available for all isoforms.

                        <br><br><b>Variant effects across the pangenome</b>: The view shows the predicted impacts across the natural variation in maize. Using B73 as a reference, all the proteins in the pan-gene set are aligned and the variations are color-coded based on potential phenotypic impact. Note: insertions and deletions are shown, but no effect score is predicted and only canonical isoforms are used for each gene model.
                        </div>
                </div>
            </form>
        </div>
    </div>

    <div id="help" class="content">
        <div class="text-content">
        <div class="gene-model-title">Maize Variant Effects Tool</div>
        <p>The maize variant effects tool offers a comprehensive exploration of the potential impacts of missense variants on maize proteins. With this tool, users can delve into four specialized views:</p>
        <ol class="help-list">
            <li><strong>Gene Summary:</strong> This section delivers a detailed look at annotations and provides valuable information regarding gene models and proteins. The gene summary includes gene and gene model annotations, protein annotations, links to 3D structure tools, and trait data from three collections of genome-wide association studies.</li>
            <li><strong>Variant Effects in B73:</strong> The first variant effects view provides a visualization of the variant effects of all possible amino acid substitutions in the B73 maize line. This section has two heatmaps, one providing a broad overview and another providing detailed insight. The width of the heatmap represents each position of the reference protein and the height of the position represents the 20 possible amino acid substitutions. The cells within the heatmap shift from blue (benign outcomes) to red (strong phenotypic impact). Hovering over a cell provides additional information about the possible substitution including position, substitution, and score. A slider bar in the center of the page controls which portion of the protein is shown on the zoomed page. Two tracks above the heat show the locations of Pfam domains and predicted secondary structures which further provide context regarding the functional and structural roles in the consequence of amino acid substitutions. </li>
            <li><strong>Variant Effects across the Pangenome:</strong> The second view also displays heat maps with a broad and detailed view, but in this instance, they show the effects of naturally occurring variations on maize proteins from across the pangenome. The width of the heatmap represents each position of the protein of the reference protein while the height of the heat map represents each protein in the pangenome aligned to the reference protein. Insertions and deletions in the alignments of the proteins are represented by a ‘-’. The heatmap representation shifts from blue (benign outcomes) to red (strong phenotypic impact) for variants within the pangenome as compared to the reference protein. Hovering over a cell provides additional information about the possible substitution including B73 position, target position, target genome, target gene model, substitution, and score. This view is only available for the canonical transcript of each gene model. </li>
            <li><strong>Search:</strong> The search section has a search bar and brief description of the tool. The search will accept gene models, transcripts, and protein identifiers.</li>
        </ol>
        <div class="gene-model-title">Data Sources:</div>
            PanEffect integrates seven different datasets to explore the potential phenotypic consequences of missense mutations. A variety of tools and datasets were used to generate input files for PanEffect. The code can be found at <a href="#" target="_blank">GitHub</a> and the data can be found at the <a href="#" target="_blank">MaizeGDB downloads</a>.
            <ul class="help-list">
                <li><strong>Variant effect scores in B73</strong>:  The variant effect scores are calculated by using the ESM protein language model (<a href="https://doi.org/10.1126/science.ade2574" target="_blank">Z. Lin et al., 2023</a>) through the <a href="https://github.com/ntranoslab/esm-variants" target="_blank">esm-variants tool</a> (<a href="https://doi.org/10.1038/s41588-023-01465-0" target="_blank">Brandes et al., 2023</a>). The esm-variant tool calculates variant effect scores based on the log-likelihood ratio difference between the variant and its wild type. Scores above -7 indicate benign outcomes, while scores below -7 suggest possible phenotypic effects (Brandes et al., 2023). A score of 0 denotes when the variant and wild-type are the same. PanEffect provides a Python script to convert the the esm-variant tool output into a CSV file containing the x and y position of the heatmap, the variant score, and the amino acid code for the wild-type (B73) and the substitution. </li>
                <li><strong>Pangenome multiple sequence alignments</strong>:  The multiple sequence alignments (MSA) were computed using the software package Pandagma which uses an all-vs-all approach to align every protein in a given pan-gene group. Pandagma generated the multiple sequence alignemnts for the  maize reference genomes in the pangenome including: three versions of B73 (<a href="https://doi.org/10.1126/science.abg5289" target="_blank">Hufford et al., 2021</a>; <a href="" target="_blank">Jiao et al., 2017</a>), a set of <a href="https://www.maizegdb.org/NAM_project" target="_blank">25 diverse maize</a> lines called the Nested Association Mapping (NAM) panel, <a href="https://doi.org/10.1038/s41588-022-01283-w" target="_blank">12 lines</a> used in Chinese breeding programs, <a href="https://www.europeanmaize.net/" target="_blank">4 European lines</a>,  <a href="https://panandropogoneae.com/" target="_blank">12 lines from the Andropogoneae tribee</a> of grasses, and a additional <a href="https://www.maizegdb.org/genome" target="_blank">set of high-qualiry maize lines</a> available at MaizeGDB. The maize genome B73 RefGen_v5 (<a href="https://www.maizegdb.org/genome/assembly/Zm-B73-REFERENCE-NAM-5.0" target="_blank">Zm-B73-REFERENCE-NAM-5.0</a>) is used as a reference and contains 39,755 gene models and 75,539 transcripts.</li>
                <li><strong>Variant effect scores in the pan-genome</strong>:  The variant effect scores for B73 were combined with the pangenome MSAs to create heatmap representations for variant effects across the maize pangenome. Using B73 as a reference, the natural variation in each protein of the pangenome were given a score using B73 as the wiltype and the target protein as the variant substitution. In <a href="" target="_blank">GitHub</a>, in <a href="" target="_blank">Github</a> PanEffect has a Python script to combine these two data types to create a TSV listing the x and y positions of the heatmap, the variant score, the positions of the amino acid in B73 and target protein, and the amino acid codes at those positions.</li>
                <li><strong>Secondary protein structures</strong>:  <a href="https://www.maizegdb.org/" target="_blank">MaizeGDB</a> has predicted the three-dimensional protein structures for all protein isoforms in the B73 genome (<a href="" target="_blank">Woodhouse et al., 2023</a>). These structures are used as input to the DSSP tool (<a href="" target="_blank">Kabsch and Sander, 1983</a>) to assign secondary structures (alpha-helices and beta sheets) to the proteins. In <a href="" target="_blank">Github</a> there is a Python script that converts the DSSP output to a TSV file listing the position, amino acid, and secondary structure code for each B73 isoform.</li>
                <li><strong>Pfam domains</strong>:  The <a href="https://download.maizegdb.org/Zm-B73-REFERENCE-NAM-5.0/" target="_blank">locations</a> of <a href="https://www.ebi.ac.uk/interpro" target="_blank">Pfam</a> functional domains for the B73 proteome were calculated by the NAM sequencing consortium using <a href="https://www.ebi.ac.uk/interpro/search/sequence/" target="_blank">Interproscan</a> and are available at MaizeGDB. The <a href="" target="_blank">Github</a> repository has a Python script that generates a separate TSV file containing the Interproscan (including PfAM position, ID, name, and Gene Ontology terms) for each B73 transcript.</li>
                <li><strong>Functional annotations</strong>:  <a href="https://www.maizegdb.org/" target="_blank">MaizeGDB</a> hosts a set of functional annotations for maize genome assemblies including <a href="https://www.uniprot.org/" target="_blank">UniProt</a> annotations, canonical gene transcript IDs, and manually annotated gene names and symbols for each gene model in B73. In GitHub, PanEffect has a Python script to parse this information into a TSV file for each B73 gene model.</li>
                <li><strong>GWAS trait annotations</strong>:  <a href="https://www.maizegdb.org/" target="_blank">MaizeGDB</a> hosts  <a href="https://ars-usda.app.box.com/v/maizegdb-public/folder/165407794668" target="_blank">three GWAS atlas datasets</a>. Each of these datasets compiled and/or generated sets of single nucleotide polymorphisms (SNPs) linked to specific traits using Genome-wide association studies (GWAS). The first set was compiled in 2014 (<a href="https://doi.org/10.1371/journal.pgen.1004845" target="_blank">Wallace et al., 2014</a>) and has GWAS mappings for over 40 traits to 40,000 SNP locations in the B73 genome. The second set is from the 2022 update of the GWAS Atlas database (<a href="https://doi.org/10.1093/nar/gkz828" target="_blank">Tian et al., 2020</a>), and combined GWAS data from 133 papers covering 531 studies for 279 traits across 42,000 SNP loci. The third dataset is from a 2022 study (<a href="https://doi.org/10.1038/s41477-022-01190-2" target="_blank">Li et al., 2022</a> that performed GWAS for 21 important agronomic traits across 1,604 inbred lines and identified 2,360 significant associations at 1,847 SNP loci. PanEffect has a Python script that finds any SNP position within 1,000 base pairs of the start and end position of a B73 gene model. The data for the three datasets are merged and a final TSV file was created for each gene model, listing the trait name and which study it came from. </li>
            </ul>
        </div>
    </div>

    <div class="footer">
        Site designed and hosted by <a target="_blank" href="https://www.maizegdb.org/">MaizeGDB</a> with support from the USDA<br>
        Powered by the PanEffect framework
    </div>
</div>
</body>

</html>
